File: decl-ts/tests/packages_test.ts
// Packages, manifests, and the lock file (§8.6–8.7) through the package
// corpus (tests/packages/cases.json): the manifest and resolution errors
// as the command line reports them, and the lock — written by the API into
// a copy of the package, bit for bit the committed text, verified clean,
// then failing closed on content drift, version drift, and a missing entry.
import { spawnSync } from 'node:child_process';
import { readFileSync, writeFileSync, appendFileSync, mkdtempSync, cpSync, rmSync } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import { initParser } from '../src/node.ts';
import { openPackageUniverse, writeLock, verifyLock } from '../src/package.ts';
import { check, total, root, cli } from './common/check.ts';

await initParser();
const cases = JSON.parse(readFileSync(join(root, 'tests/packages/cases.json'), 'utf8'));

/** the codes `decl check` reports for an entry, in order */
const codesOf = (entry: string): string[] => {
  const r = spawnSync(process.execPath, [cli, 'check', entry], { encoding: 'utf8', cwd: root });
  return [...r.stderr.matchAll(/\[(E\d{4})\]/g)].map((m) => m[1]);
};

console.log('== packages: manifest and resolution errors ==');
for (const e of cases.errors)
  check(
    `${e.entry}: ${e.codes.join(', ')}`,
    JSON.stringify(codesOf(e.entry)) === JSON.stringify(e.codes),
    JSON.stringify(codesOf(e.entry)),
  );

console.log('== packages: the lock file, reproducible and fail-closed ==');
{
  const lock = cases.lock;
  const expected = readFileSync(join(root, lock.lock), 'utf8');
  // a fresh copy of the package, with its lock written by the API
  const fresh = (): { dir: string; entry: string } => {
    const dir = mkdtempSync(join(tmpdir(), 'decl-pkg-'));
    cpSync(join(root, lock.package), dir, { recursive: true });
    const entry = join(dir, lock.entry);
    writeLock(openPackageUniverse(entry)!);
    return { dir, entry };
  };
  const { dir, entry } = fresh();
  check(
    'the lock is the committed text, byte for byte',
    readFileSync(join(dir, 'decl.lock'), 'utf8') === expected,
  );
  check('a fresh lock verifies clean', verifyLock(openPackageUniverse(entry)!).length === 0);
  check('the command line accepts the locked package', codesOf(entry).length === 0);
  rmSync(dir, { recursive: true, force: true });
  for (const d of lock.drift) {
    const { dir, entry } = fresh();
    for (const [file, text] of Object.entries(d.append ?? {}))
      appendFileSync(join(dir, file), text as string);
    if (d.lock_replace)
      writeFileSync(
        join(dir, 'decl.lock'),
        readFileSync(join(dir, 'decl.lock'), 'utf8').replace(d.lock_replace[0], d.lock_replace[1]),
      );
    if (d.lock_text !== undefined) writeFileSync(join(dir, 'decl.lock'), d.lock_text);
    const got = codesOf(entry);
    check(d.name, JSON.stringify(got) === JSON.stringify(d.codes), JSON.stringify(got));
    rmSync(dir, { recursive: true, force: true });
  }
}
total();

File: decl-rs/tests/packages_test.rs
//! Packages, manifests, and the lock file (§8.6–8.7) through the package
//! corpus (tests/packages/cases.json): the manifest and resolution errors
//! as the command line reports them, and the lock — written by the API into
//! a copy of the package, bit for bit the committed text, verified clean,
//! then failing closed on content drift, version drift, and a missing entry.
mod common;
use common::*;
use decl_lang::package::{open_package_universe, verify_lock, write_lock};
use std::path::PathBuf;

#[test]
fn packages_corpus() {
    let root = root();
    let text = std::fs::read_to_string(root.join("tests/packages/cases.json")).unwrap();
    let cases = parse(&text);
    let mut failures: Vec<String> = vec![];
    let codes_of = |v: &Value| -> Vec<String> {
        match v {
            Value::JArr(xs) => xs.iter().map(|x| common::text(x).to_string()).collect(),
            _ => vec![],
        }
    };

    // the manifest and resolution errors, as `decl check` reports them
    if let Some(Value::JArr(errors)) = get(&cases, "errors") {
        for e in errors.iter() {
            let entry = common::text(get(e, "entry").unwrap());
            let want = codes_of(get(e, "codes").unwrap());
            let got = check_codes(&root.join(entry));
            if got != want {
                failures.push(format!("{entry}: expected {want:?}, got {got:?}"));
            }
        }
    }

    // the lock: reproducible, then fail-closed
    let lock = get(&cases, "lock").unwrap();
    let package = root.join(common::text(get(lock, "package").unwrap()));
    let entry_name = common::text(get(lock, "entry").unwrap()).to_string();
    let expected =
        std::fs::read_to_string(root.join(common::text(get(lock, "lock").unwrap()))).unwrap();
    // a fresh copy of the package, with its lock written by the API
    let fresh = || -> (PathBuf, PathBuf) {
        let dir = temp_dir("pkg");
        copy_dir(&package, &dir);
        let entry = dir.join(&entry_name);
        write_lock(&open_package_universe(&entry).expect("in a package"));
        (dir, entry)
    };
    let (dir, entry) = fresh();
    let written = std::fs::read_to_string(dir.join("decl.lock")).unwrap();
    if written != expected {
        failures.push(format!("the lock is not the committed text: {written:?}"));
    }
    if !verify_lock(&open_package_universe(&entry).unwrap()).is_empty() {
        failures.push("a fresh lock does not verify clean".into());
    }
    let codes = check_codes(&entry);
    if !codes.is_empty() {
        failures.push(format!(
            "the command line refuses the locked package: {codes:?}"
        ));
    }
    let _ = std::fs::remove_dir_all(&dir);
    if let Some(Value::JArr(drifts)) = get(lock, "drift") {
        for d in drifts.iter() {
            let name = common::text(get(d, "name").unwrap());
            let (dir, entry) = fresh();
            if let Some(Value::JObj(files)) = get(d, "append") {
                for (f, t) in files.iter() {
                    let p = dir.join(f);
                    let mut s = std::fs::read_to_string(&p).unwrap();
                    s.push_str(common::text(t));
                    std::fs::write(&p, s).unwrap();
                }
            }
            if let Some(Value::JArr(pair)) = get(d, "lock_replace") {
                let p = dir.join("decl.lock");
                let s = std::fs::read_to_string(&p).unwrap();
                std::fs::write(
                    &p,
                    s.replacen(common::text(&pair[0]), common::text(&pair[1]), 1),
                )
                .unwrap();
            }
            if let Some(t) = get(d, "lock_text") {
                std::fs::write(dir.join("decl.lock"), common::text(t)).unwrap();
            }
            let want = codes_of(get(d, "codes").unwrap());
            let got = check_codes(&entry);
            if got != want {
                failures.push(format!("{name}: expected {want:?}, got {got:?}"));
            }
            let _ = std::fs::remove_dir_all(&dir);
        }
    }
    assert!(
        failures.is_empty(),
        "packages corpus failures:\n  {}",
        failures.join("\n  ")
    );
}
